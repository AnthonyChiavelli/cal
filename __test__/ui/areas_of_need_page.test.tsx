import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import AreasOfNeedPage from "@/app/components/areas_of_need_page";

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
  usePathname() {
    return {};
  },
}));

// Monkeypatch
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

const mockAreasOfNeed = [
  {
    id: "1",
    createdAt: new Date("2021-02-02"),
    updatedAt: new Date("2021-02-02"),
    name: "Charcuterie",
  },
  {
    id: "2",
    createdAt: new Date("2021-04-04"),
    updatedAt: new Date("2021-03-03"),
    name: "Ham Slicing",
  },
];

describe("AreasOfNeedPage", () => {
  it("Renders a consistent snapshot when empty", async () => {
    const container = render(<AreasOfNeedPage areasOfNeed={[]} />);
    expect(container).toMatchSnapshot();
  });

  it("Renders a consistent snapshot when filled with data", async () => {
    const container = render(<AreasOfNeedPage areasOfNeed={mockAreasOfNeed} />);
    expect(container).toMatchSnapshot();
  });

  it("Renders correctly when empty", async () => {
    render(<AreasOfNeedPage areasOfNeed={[]} />);

    expect(screen.getByText("No areas of need yet!")).toBeInTheDocument();
  });

  it("Renders correctly when filled with data", async () => {
    render(<AreasOfNeedPage areasOfNeed={mockAreasOfNeed} />);
    expect(screen.queryByText("No areas of need yet!")).toBeNull();
    for (const area of mockAreasOfNeed) {
      expect(screen.getByText(area.name)).toBeInTheDocument();
      expect(screen.getByText(area.createdAt.toLocaleString())).toBeInTheDocument();
    }
  });

  it("Should show a deletion modal on delete and call the delete method on confirm", async () => {
    const mockDelete = jest.fn();
    render(<AreasOfNeedPage areasOfNeed={mockAreasOfNeed} deleteAreaOfNeed={mockDelete} />);
    window.ResizeObserver = ResizeObserver;

    fireEvent.click(screen.getAllByText("Delete")[0]);
    expect(screen.getByText("Are you sure you want to delete 'Charcuterie'?")).toBeInTheDocument();
    fireEvent.click(screen.getAllByText("Okay")[0]);

    expect(mockDelete).toHaveBeenCalledTimes(1);
    expect(mockDelete).toHaveBeenCalledWith("1");
  });
});

describe("AddAreaOfNeedModal", () => {
  it("Show and properly handle input", async () => {
    const mockOnCreate = jest.fn();
    render(<AreasOfNeedPage areasOfNeed={mockAreasOfNeed} updateOrCreateAreaOfNeed={mockOnCreate} />);

    // Not implemented in jest, need to monkeypatch
    // @ts-ignore
    window.ResizeObserver = ResizeObserver;
    fireEvent.click(screen.getByText("Add Area of Need"));

    expect(screen.getByText("Create new 'Area of need'")).toBeInTheDocument();
    expect(screen.getByTestId("area-of-need-name")).toBeInTheDocument();

    expect(screen.getByText("Create")).toBeDisabled();
    fireEvent.change(screen.getByTestId("area-of-need-name"), { target: { value: "Test" } });
    expect(screen.getByText("Create")).not.toBeDisabled();

    fireEvent.click(screen.getByText("Create"));
    expect(mockOnCreate).toHaveBeenCalledTimes(1);
    expect(mockOnCreate).toHaveBeenCalledWith("Test");
  });

  it("Blocks creation of duplicate-named entity, with case-insensivity", async () => {
    render(<AreasOfNeedPage areasOfNeed={mockAreasOfNeed} />);

    // Not implemented in jest, need to monkeypatch
    // @ts-ignore
    window.ResizeObserver = ResizeObserver;
    fireEvent.click(screen.getByText("Add Area of Need"));

    expect(screen.getByText("Create new 'Area of need'")).toBeInTheDocument();
    expect(screen.getByTestId("area-of-need-name")).toBeInTheDocument();
    fireEvent.change(screen.getByTestId("area-of-need-name"), { target: { value: "Charcuterie" } });
    expect(screen.getByText("Create")).toBeDisabled();

    fireEvent.change(screen.getByTestId("area-of-need-name"), { target: { value: "chArcuTerIe" } });
    expect(screen.getByText("Create")).toBeDisabled();
  });
});
