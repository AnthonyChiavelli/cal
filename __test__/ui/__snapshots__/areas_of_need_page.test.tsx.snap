// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreasOfNeedPage Renders a consistent snapshot when empty 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <section
          class="mt-5"
        >
          <h1
            class=""
          >
            Areas of Need
          </h1>
          <table
            class="min-w-full divide-y divide-gray-300"
          >
            <thead>
              <tr
                class="text-sm text-left font-semibold bg-slate-400 text-white"
              >
                <th
                  class="py-3 px-4"
                >
                  Area
                </th>
                <th
                  class="py-3 px-4"
                >
                  Created At
                </th>
                <th
                  class="py-3 px-4"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="bg-white"
            >
              <tr
                class="bg-white italic"
              >
                <td
                  class="text-center py-3"
                  colspan="4"
                >
                  No areas of need yet!
                </td>
              </tr>
            </tbody>
          </table>
        </section>
        <section
          class="mt-5"
        >
          <button
            class="text-sm font-semibold px-3 py-2 flex flex-row justify-center rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-indigo-600 hover:bg-indigo-500  text-white"
            data-cy="button-add-area-of-need"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 mr-2"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 4.5v15m7.5-7.5h-15"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span>
              Add Area of Need
            </span>
          </button>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <section
        class="mt-5"
      >
        <h1
          class=""
        >
          Areas of Need
        </h1>
        <table
          class="min-w-full divide-y divide-gray-300"
        >
          <thead>
            <tr
              class="text-sm text-left font-semibold bg-slate-400 text-white"
            >
              <th
                class="py-3 px-4"
              >
                Area
              </th>
              <th
                class="py-3 px-4"
              >
                Created At
              </th>
              <th
                class="py-3 px-4"
              >
                Actions
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white"
          >
            <tr
              class="bg-white italic"
            >
              <td
                class="text-center py-3"
                colspan="4"
              >
                No areas of need yet!
              </td>
            </tr>
          </tbody>
        </table>
      </section>
      <section
        class="mt-5"
      >
        <button
          class="text-sm font-semibold px-3 py-2 flex flex-row justify-center rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-indigo-600 hover:bg-indigo-500  text-white"
          data-cy="button-add-area-of-need"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 mr-2"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4.5v15m7.5-7.5h-15"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span>
            Add Area of Need
          </span>
        </button>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AreasOfNeedPage Renders a consistent snapshot when filled with data 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <section
          class="mt-5"
        >
          <h1
            class=""
          >
            Areas of Need
          </h1>
          <table
            class="min-w-full divide-y divide-gray-300"
          >
            <thead>
              <tr
                class="text-sm text-left font-semibold bg-slate-400 text-white"
              >
                <th
                  class="py-3 px-4"
                >
                  Area
                </th>
                <th
                  class="py-3 px-4"
                >
                  Created At
                </th>
                <th
                  class="py-3 px-4"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="bg-white"
            >
              <tr
                class="text-sm text-gray-800 even:bg-gray-50"
              >
                <td
                  class="py-3 px-4"
                >
                  Charcuterie
                </td>
                <td
                  class="py-3 px-4"
                >
                  2/1/2021, 7:00:00 PM
                </td>
                <td
                  class="py-3 px-4 gap-2 flex flex-row"
                >
                  <button
                    class="text-red-500 hover:text-red-700"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr
                class="text-sm text-gray-800 even:bg-gray-50"
              >
                <td
                  class="py-3 px-4"
                >
                  Ham Slicing
                </td>
                <td
                  class="py-3 px-4"
                >
                  4/3/2021, 8:00:00 PM
                </td>
                <td
                  class="py-3 px-4 gap-2 flex flex-row"
                >
                  <button
                    class="text-red-500 hover:text-red-700"
                  >
                    Delete
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </section>
        <section
          class="mt-5"
        >
          <button
            class="text-sm font-semibold px-3 py-2 flex flex-row justify-center rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-indigo-600 hover:bg-indigo-500  text-white"
            data-cy="button-add-area-of-need"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 mr-2"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 4.5v15m7.5-7.5h-15"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span>
              Add Area of Need
            </span>
          </button>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <section
        class="mt-5"
      >
        <h1
          class=""
        >
          Areas of Need
        </h1>
        <table
          class="min-w-full divide-y divide-gray-300"
        >
          <thead>
            <tr
              class="text-sm text-left font-semibold bg-slate-400 text-white"
            >
              <th
                class="py-3 px-4"
              >
                Area
              </th>
              <th
                class="py-3 px-4"
              >
                Created At
              </th>
              <th
                class="py-3 px-4"
              >
                Actions
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white"
          >
            <tr
              class="text-sm text-gray-800 even:bg-gray-50"
            >
              <td
                class="py-3 px-4"
              >
                Charcuterie
              </td>
              <td
                class="py-3 px-4"
              >
                2/1/2021, 7:00:00 PM
              </td>
              <td
                class="py-3 px-4 gap-2 flex flex-row"
              >
                <button
                  class="text-red-500 hover:text-red-700"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr
              class="text-sm text-gray-800 even:bg-gray-50"
            >
              <td
                class="py-3 px-4"
              >
                Ham Slicing
              </td>
              <td
                class="py-3 px-4"
              >
                4/3/2021, 8:00:00 PM
              </td>
              <td
                class="py-3 px-4 gap-2 flex flex-row"
              >
                <button
                  class="text-red-500 hover:text-red-700"
                >
                  Delete
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </section>
      <section
        class="mt-5"
      >
        <button
          class="text-sm font-semibold px-3 py-2 flex flex-row justify-center rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-indigo-600 hover:bg-indigo-500  text-white"
          data-cy="button-add-area-of-need"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 mr-2"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4.5v15m7.5-7.5h-15"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span>
            Add Area of Need
          </span>
        </button>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
